contracts/integrations/test/exchange/transaction_protocol_fee_test.ts

```diff
@@ -0,0 +1,500 @@
+                const txReceipt = await deployment.exchange
+                    .executeTransaction(recursiveTransaction, recursiveTransaction.signature)
+                    .awaitTransactionSuccessAsync({ from: alice.address, value: REFUND_AMOUNT });
+                expectedBalances.simulateFills([order], alice.address, txReceipt, deployment, REFUND_AMOUNT);
+            });
+            it('Alice executeTransaction => Alice executeTransaction => Bob fillOrder; protocol fee in wETH', async () => {
+                const recursiveData = deployment.exchange
+                    .executeTransaction(transactionB, transactionB.signature)
+                    .getABIEncodedTransactionData();
+                const recursiveTransaction = await alice.signTransactionAsync({ data: recursiveData });
+                const txReceipt = await deployment.exchange
+                    .executeTransaction(recursiveTransaction, recursiveTransaction.signature)
+                    .awaitTransactionSuccessAsync({ from: alice.address, value: REFUND_AMOUNT });
+                expectedBalances.simulateFills([order], bob.address, txReceipt, deployment, REFUND_AMOUNT);
+            });
+            it('Alice executeTransaction => Alice executeTransaction => wETH-less taker fillOrder; reverts because protocol fee cannot be paid', async () => {
+                const data = exchangeDataEncoder.encodeOrdersToExchangeData(ExchangeFunctionName.FillOrder, [order]);
+                const transaction = await wethless.signTransactionAsync({ data });
+                const recursiveData = deployment.exchange
+                    .executeTransaction(transaction, transaction.signature)
+                    .getABIEncodedTransactionData();
+                const recursiveTransaction = await alice.signTransactionAsync({ data: recursiveData });
+                const tx = deployment.exchange
+                    .executeTransaction(recursiveTransaction, recursiveTransaction.signature)
+                    .awaitTransactionSuccessAsync({ from: alice.address, value: REFUND_AMOUNT });
+                const expectedError = new ExchangeRevertErrors.TransactionExecutionError(
+                    transactionHashUtils.getTransactionHashHex(recursiveTransaction),
+                    protocolFeeError(order, transaction).encode(),
+                );
+                return expect(tx).to.revertWith(expectedError);
+            });
+        });
+    });
+    describe('batchExecuteTransactions', () => {
+        let expectedBalances: LocalBalanceStore;
+        beforeEach(async () => {
+            await balanceStore.updateBalancesAsync();
+            expectedBalances = LocalBalanceStore.create(balanceStore);
+        });
+        afterEach(async () => {
+            await balanceStore.updateBalancesAsync();
+            balanceStore.assertEquals(expectedBalances);
+        });
+
+        describe('Simple', () => {
```

Comment:
Seems like a lot of these could just be generated combinations? Might not be worth the effort rn though.